bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+		/// Convert a command from Ethereum to an XCM message.
+		type MessageConverter: ConvertMessage;
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self>;
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was received from Ethereum
+		MessageReceived {
+			/// The message nonce
+			nonce: u64,
+			/// ID of the XCM message which was forwarded to the final destination parachain
+			message_id: [u8; 32],
+		},
+		/// Set OperatingMode
+		OperatingModeChanged { mode: BasicOperatingMode },
+		/// XCM delivery fees were paid.
+		FeesPaid { paying: Location, fees: Assets },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Message came from an invalid outbound channel on the Ethereum side.
+		InvalidGateway,
+		/// Account could not be converted to bytes
+		InvalidAccount,
+		/// Message has an invalid envelope.
+		InvalidMessage,
+		/// Message has an unexpected nonce.
+		InvalidNonce,
+		/// Fee provided is invalid.
+		InvalidFee,
+		/// Message has an invalid payload.
+		InvalidPayload,
+		/// Message channel is invalid
+		InvalidChannel,
+		/// The max nonce for the type has been reached
+		MaxNonceReached,
+		/// Cannot convert location
+		InvalidAccountConversion,
+		/// Pallet is halted
+		Halted,
+		/// The operation required fees to be paid which the initiator could not meet.
+		FeesNotMet,
+		/// The desired destination was unreachable, generally because there is a no way of routing
+		/// to it.
+		Unreachable,
+		/// There was some other issue (i.e. not to do with routing) in sending the message.
+		/// Perhaps a lack of space for buffering the message.
+		SendFailure,
+		/// Invalid foreign ERC-20 token ID
+		InvalidAsset,
+		/// Cannot reachor a foreign ERC-20 asset location.
+		CannotReanchor,
+		/// Reward payment Failure
+		RewardPaymentFailed,
+		/// Message verification error
+		Verification(VerificationError),
+
+	}
+
+	impl<T: Config> From<SendError> for Error<T> {
+		fn from(e: SendError) -> Self {
+			match e {
+				SendError::Fees => Error::<T>::FeesNotMet,
+				SendError::NotApplicable => Error::<T>::Unreachable,
+				_ => Error::<T>::SendFailure,
+			}
+		}
+	}
+
+	impl<T: Config> From<ConvertMessageError> for Error<T> {
+		fn from(e: ConvertMessageError) -> Self {
+			match e {
+				ConvertMessageError::InvalidAsset => Error::<T>::InvalidAsset,
+				ConvertMessageError::CannotReanchor => Error::<T>::CannotReanchor,
+			}
+		}
+	}
+
+	/// The nonce of the message been processed or not
+	#[pallet::storage]
+	pub type NonceBitmap<T: Config> = StorageMap<_, Twox64Concat, u128, u128, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> where Location: From<<T as frame_system::Config>::AccountId> {
```

Comment:
I think we may need an explict converter from AccountId to Location.

Can be injected from runtime to avoid binding to `AccountId32`, https://github.com/vgeddes/polkadot-sdk/blob/7b96c918711ee830c26cd9944106fe1ad40df1b1/polkadot/runtime/rococo/src/lib.rs#L1138 FYI

Reply:
Why do you think we need an explicit converter? 🤔 https://github.com/paritytech/polkadot-sdk/pull/6697#discussion_r1935431301

Reply:
IIUC It's a converter from `[u8; 32]` to Location,  but the `T::AccountId` here is a type inferred from the Runtime,  though it's `[u8;32]` on BH, we may not want to narrow it down on Pallet definition.

I made some change in https://github.com/vgeddes/polkadot-sdk/pull/3 to make it compile, mainly in this [commit](https://github.com/vgeddes/polkadot-sdk/pull/3/commits/9d7f35650f289788e287d06e657072608d7e6a2f)

Please check if that make sense.

Reply:
Makes sense! I agree that seems better.